source/fuzz/transformation_outline_selection_construct.cpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_outline_selection_construct.h"
+
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationOutlineSelectionConstruct::
+    TransformationOutlineSelectionConstruct(
+        const protobufs::TransformationOutlineSelectionConstruct& message)
+    : message_(message) {}
+
+TransformationOutlineSelectionConstruct::
+    TransformationOutlineSelectionConstruct(uint32_t new_header_block_id,
+                                            uint32_t new_merge_block_id) {
+  message_.set_new_header_block_id(new_header_block_id);
+  message_.set_new_merge_block_id(new_merge_block_id);
+}
+
+bool TransformationOutlineSelectionConstruct::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // Check that it is possible to outline a region of blocks without breaking
+  // domination and structured control flow rules.
+  if (!IsApplicableToBlockRange(ir_context, message_.new_header_block_id(),
+                                message_.new_merge_block_id())) {
+    return false;
+  }
+
+  // There must exist an irrelevant boolean constant to be used as a condition
+  // in the OpBranchConditional instruction.
+  return fuzzerutil::MaybeGetBoolConstant(ir_context, transformation_context,
+                                          true, true) != 0;
+}
+
+void TransformationOutlineSelectionConstruct::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  auto* new_header_block =
+      ir_context->cfg()->block(message_.new_header_block_id());
+  assert(new_header_block->terminator()->opcode() == SpvOpBranch &&
+         "Should condition should've been checked in the IsApplicable");
+
+  const auto successor_id =
+      new_header_block->terminator()->GetSingleWordInOperand(0);
+
+  // Change |entry_block|'s terminator to |OpBranchConditional|.
+  new_header_block->terminator()->SetOpcode(SpvOpBranchConditional);
+  new_header_block->terminator()->SetInOperands(
+      {{SPV_OPERAND_TYPE_ID,
+        {fuzzerutil::MaybeGetBoolConstant(ir_context, *transformation_context,
+                                          true, true)}},
```

Comment:
For diversity it would be nice if the transformation could specify whether the condition needs to be true or false, with the branch targets being set up correspondingly.

Reply:
Fixed in 44c64b8.